File: component/src/main/java/com/mackwu/component/core/pattern/factory/ConcreteFactory.java
package com.mackwu.component.core.pattern.factory;

import java.lang.reflect.Constructor;

/**
 * ===================================================
 * Created by MackWu on 2021/4/6 11:38
 * <a href="mailto:wumengjiao828@163.com">Contact me</a>
 * <a href="https://github.com/mackwu828">Follow me</a>
 * ===================================================
 */
public class ConcreteFactory extends Factory {

    @Override
    public <T extends Product> T createProduct(Class<T> c) {
        T product = null;
        try {
            Constructor<T> constructor = c.getConstructor();
            product = constructor.newInstance();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return product;
    }
}

File: java_kt/src/main/java/com/mackwu/java_kt/kt/pattern/factory/ConcreteFactory.kt
package com.mackwu.java_kt.kt.pattern.factory

/**
 * ===================================================
 * Created by MackWu on 2020/1/3 1:21
 * <a href="mailto:wumengjiao828@163.com">Contact me</a>
 * <a href="https://github.com/mackwu828">Follow me</a>
 * ===================================================
 * 具体工厂类
 */
class ConcreteFactory : Factory{

    override fun <T : Product> createProduct(productClass: Class<T>): T{
        return Class.forName(productClass.name).newInstance() as T
    }
}

fun main() {
    val concreteFactory = ConcreteFactory()
    val productA = concreteFactory.createProduct(ProductA::class.java)
    val productB = concreteFactory.createProduct(ProductB::class.java)
}